src/xmipp/libraries/tomo/tomo_ctf_wiener2d_correction.cpp

```diff
@@ -0,0 +1,203 @@
+ *
+ * Authors:     Jose Luis Vilas (jlvilas@cnb.csic.es)
+ *
+ * Spanish Research Council for Biotechnology, Madrid, Spain
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ * 02111-1307  USA
+ *
+ *  All comments concerning this program package may be sent to the
+ *  e-mail address 'xmipp@cnb.csic.es'
+ ***************************************************************************/
+
+#include "tomo_ctf_wiener2d_correction.h"
+#include <sys/stat.h>
+#include "data/ctf.h"
+#include <core/metadata_extension.h>
+#include "reconstruction/ctf_correct_wiener2d.h"
+#include "core/transformations.h"
+#include <limits>
+#include <type_traits>
+
+
+void ProgCTFWiener2DCorrection::defineParams()
+{
+	addUsageLine("This method performs a CTF correction of a tilt series");
+	addParamsLine("   -i <input_file>                    : Metadata with the tilt series");
+	addParamsLine("   -o <output_volume>          	 	 : Output volume.");
+	addParamsLine("   --sampling <sampling=1>            : Pixel size (Angstrom)");
+	addParamsLine("   [--wiener_constant <wc=0.1>        : Wiener constant");
+	addParamsLine("   [--threads <nthreads=1>]           : (Optional) Number of threads to be used");
+
+	addExampleLine("Obtain the average of a set of subtomograms from a metadata file", false);
+	addExampleLine("xmipp_tomo_average_subtomograms -i inputSubtomos.xmd -o average.mrc");
+}
+
+void ProgCTFWiener2DCorrection::readParams()
+{
+	fnIn = getParam("-i");
+	fnOut = getParam("-o");
+	sampling = getDoubleParam("--sampling");
+	wc = getDoubleParam("--wiener_constant");
+	sigmaDf = getDoubleParam("--sigmaDf");
+	nthreads = getIntParam("--threads");
+}
+
+void ProgCTFWiener2DCorrection::gaussianMask(MultidimArray<double> &cumMask, 
+											MultidimArray<double> &tiMask,  
+											MultidimArray<double> &ptrImg, int x0, int stripeSize)
+{
+	auto xdim = XSIZE(ptrImg);
+	auto ydim = XSIZE(ptrImg);
+
+	// To set sigma: We assume that the gaussian g(stripeSize) = 0.1
+	// 0.1 = exp(-(stripeSize)^2/(2*sigma2)),  - log 10 = - (stripeSize)^2/(2*sigma2)
+	// therefore sigma2 = stripeSize^2/(2*log 10)
+	double sigma2 = (double) stripeSize*stripeSize/(log(100.0));
+
+	long n = 0;
+	for (size_t i = 0; i<ydim; i++)
+	{
+		for (size_t j = 0; j<xdim; j++)
+		{
+			double p = (j-x0)^2;
```

Comment:
`^` means bitwise XOR in C/C++, not exponentiation. Doing a bitwise XOR with 2 leads to flipping the second bit, which is equivalent to adding or subtracting 2 depending on whether that bit is set or not. I'm quite sure you did not meant to do that. Also, be aware that in the next line you are also squaring this number. Yo probably want to keep only one of them.